src/core/meshcop/secure_transport.cpp

```diff
@@ -1052,6 +1057,7 @@ void SecureTransport::HandleTimer(void)
             mTimer.Stop();
         }
         mConnectedCallback.InvokeIfSet(mConnectEvent);
+        mConnectEvent = kInit;
```

Comment:
Another potential approach for this (might be a bigger change though)

- We can consider expanding the `State` and have it track information including the specific disconnect reason.
- This way we can fully remove `mConnectEvent` and will have all info tracked by one `mState` variable.

Today we have `kStateCloseNotify` which is used to indicate that we are disconnecting and should notify and close. We can expand this into multiple states to track the reason for disconnect/close. In `HandleTimer()` we can then map the `mState` to the proper `ConnectEvent` when we invoke the callback.

For example, we can update `State` to something like this:

```c++
    enum State : uint8_t
    {
        kStateClosed,                 // UDP socket is closed.
        kStateOpen,                   // UDP socket is open.
        kStateInitializing,           // The service is initializing.
        kStateConnecting,             // The service is establishing a connection.
        kStateConnected,              // The service has a connection established.
        kStateCloseNotifyPeerClosed,  // Disconnecting, peer closed.
        kStateCloseNotifyLocalClosed, // Disconnecting, local closed.
        kStateCloseNotifyMaxAttempts, // Closing connection, max number of attempts reached.
        kStateCloseNotifyError,       // Disconnecting, other errors.
    };
```

The approach will be similar to previous one. 
- We can add a new `Disconnect(State aNewState)` method where instead of passing the event, we pass the desired new state (one of the `kStateCloseNotify{Event}` ones).
- Public `Disconnect(void)` calls `Disconnect(kStateCloseNotifyLocalClosed)`.
- In `Process()` when we determine the `shouldDisconnect` we also determine what next state we should go to.

Thoughts?

Reply:
I think merging the mConnectEvent into mState would be nice to keep the code easier for maintenance. This is a bigger change, maybe we can consider it as a followup improvement, after we applied the simpler approach to unblock #10655 which gives a better coverage?

Reply:
Sounds good, thanks.

The approach of combing event and state might not actually simplify the code much, as we'll  need to map the state back to a `ConnectEvent` when invoking the callback.